Code/GraphMol/SynthonSpaceSearch/SynthonSpace.cpp

```diff
@@ -0,0 +1,719 @@
+              [&](int prevRes, const boost::dynamic_bitset<> &s2) {
+                return prevRes * s2.count();
+              });
+          if (numHits) {
+            results.push_back(
+                SynthonSpaceHitSet{reaction->id(), theseSynthons, numHits});
+          }
+        }
+      }
+    }
+  }
+  return results;
+}
+
+void SynthonSpace::readTextFile(const std::string &inFilename) {
+  d_fileName = inFilename;
+  std::ifstream ifs(d_fileName);
+  if (!ifs.is_open() || ifs.bad()) {
+    throw std::runtime_error("Couldn't open file " + d_fileName);
+  }
+  std::regex regexz("[\\s,]+");
+
+  std::vector<std::vector<std::string>> firstLineOpts{
+      std::vector<std::string>{"SMILES", "synton_id", "synton#", "reaction_id"},
+      std::vector<std::string>{"SMILES", "synton_id", "synton#", "reaction_id",
+                               "release"},
+      std::vector<std::string>{"SMILES", "synton_id", "synton_role",
+                               "reaction_id"}};
+  int format = -1;
+  std::string nextLine;
+  int lineNum = 1;
+
+  while (getline(ifs, nextLine)) {
+    ++lineNum;
+    if (nextLine.empty() || nextLine[0] == '#') {
+      continue;
+    }
+    if (format == -1) {
+      auto lineParts = splitLine(nextLine, regexz);
+      for (size_t i = 0; i < firstLineOpts.size(); ++i) {
+        if (lineParts == firstLineOpts[i]) {
+          format = i;
+        }
+      }
+      if (format == -1) {
+        throw std::runtime_error("Bad format for SynthonSpace file " +
+                                 d_fileName);
+      }
+      continue;
+    }
+    auto nextSynthon = splitLine(nextLine, regexz);
+    if (nextSynthon.size() < 4) {
+      throw std::runtime_error("Bad format for SynthonSpace file " +
+                               d_fileName + " on line " +
+                               std::to_string(lineNum));
+    }
+    if (auto it = d_reactions.find(nextSynthon[3]); it == d_reactions.end()) {
+      d_reactions.insert(std::make_pair(
+          nextSynthon[3], std::make_unique<SynthonSet>(nextSynthon[3])));
+    }
+    fixConnectors(nextSynthon[0]);
+    auto &currReaction = d_reactions[nextSynthon[3]];
+    size_t synthonNum{std::numeric_limits<size_t>::max()};
+    if (format == 0 || format == 1) {
+      synthonNum = std::stoi(nextSynthon[2]);
+    } else if (format == 2) {
+      // in this case it's a string "synton_2" etc.
+      synthonNum = std::stoi(nextSynthon[2].substr(7));
+    }
+    currReaction->addSynthon(synthonNum, std::make_unique<Synthon>(Synthon(
+                                             nextSynthon[0], nextSynthon[1])));
+  }
+
+  // Do some final processing.
+  for (auto &it : d_reactions) {
+    auto &reaction = it.second;
+    reaction->buildConnectorRegions();
+    reaction->assignConnectorsUsed();
+  }
+}
+
+void SynthonSpace::writeDBFile(const std::string &outFilename) const {
+  std::ofstream os(outFilename, std::fstream::binary | std::fstream::trunc);
```

Comment:
Dealing with binary save files in a cross-platform way that is future proof is a pain.

I really strongly suggest to start the db files with a check byte (to verify that it's the right type of file and the endian-ness of the binary) and a version number. Take a look at MolPickler.cpp for details. There's a bunch of code (or patterns) in there that may be helpful.

An alternative to avoid all the complication would be to use boost::serialize to handle serialization/deserialization; it takes care of most/all of the complexity for you.

Reply:
A brief look at the code looks like it is ok.  It serializes strings and pickled molecules which already take care of the endianness.

I do think a version number would be useful for future proofing and this might require endianness.   The boost serialization of this would also be fairly trivial.

Reply:
It is also serializing ints

Reply:
Yeah, I'm aware of the issues sharing binary data.  I thought that by using code from RDKit serialisers it would all be taken care of.  A version number certainly makes sense.